read.go

```diff
@@ -18,47 +18,66 @@ import (
 	"encoding/binary"
 )
 
-func (s *Segment) getDocStoredMetaAndUnCompressed(docNum uint64) (meta, data []byte, err error) {
-	_, storedOffset, n, metaLen, dataLen, err := s.getDocStoredOffsets(docNum)
-	if err != nil {
-		return nil, nil, err
+func (s *Segment) initDecompressedStoredFieldChunks(n int) {
+	s.m.Lock()
+	s.decompressedStoredFieldChunks = make(map[uint32]*segmentCacheData, n)
+	for i := uint32(0); i < uint32(n); i++ {
+		s.decompressedStoredFieldChunks[i] = &segmentCacheData{}
 	}
-
-	meta = s.storedFieldChunkUncompressed[int(storedOffset+n):int(storedOffset+n+metaLen)]
-	data = s.storedFieldChunkUncompressed[int(storedOffset+n+metaLen):int(storedOffset+n+metaLen+dataLen)]
-	return meta, data, nil
+	s.m.Unlock()
 }
 
-func (s *Segment) getDocStoredOffsets(docNum uint64) (indexOffset, storedOffset, n, metaLen, dataLen uint64, err error) {
-	indexOffset, storedOffset, err = s.getDocStoredOffsetsOnly(docNum)
+func (s *Segment) getDocStoredMetaAndUnCompressed(docNum uint64) (meta, data []byte, err error) {
+	_, storedOffset, err := s.getDocStoredOffsetsOnly(docNum)
 	if err != nil {
-		return 0, 0, 0, 0, 0, err
+		return nil, nil, err
 	}
 
 	// document chunk coder
-	chunkI := docNum / uint64(defaultDocumentChunkSize)
-	chunkOffsetStart := s.storedFieldChunkOffsets[int(chunkI)]
-	chunkOffsetEnd := s.storedFieldChunkOffsets[int(chunkI)+1]
-	compressed, err := s.data.Read(int(chunkOffsetStart), int(chunkOffsetEnd))
-	if err != nil {
-		return 0, 0, 0, 0, 0, err
+	var uncompressed []byte
+	chunkI := uint32(docNum) / defaultDocumentChunkSize
+	storedFieldDecompressed := s.decompressedStoredFieldChunks[chunkI]
+	storedFieldDecompressed.m.Lock()
+	if storedFieldDecompressed.data == nil {
```

Comment:
This section tries to reduce the amount of repeated work by doing it inside a locked section, have you considered performing the decompression outside the lock and then lock->check->update.

So we might end up with repeated work between multiple requests but less contention between them.

what do you thing? (and also @mschoch)

Reply:
We don't want decompress all chunks, just want decompress the chunk which visited. so we decompress when visit, and to avoid decompress multiple, so we use lock to ensure just do once. 

And about you talked i am not sure how to implement it, i think you can try to create a new PR or give some code for that.